### ISSUE TITLE
The "tabindex" attribute is missing from all components

### ISSUE THREAD

**Issue by tomkroschwald** [telekom/scale#668]
To ensure accessibility for components such as "scale-text-field", "scale-button" etc., we need the tabindex attribute. Unfortunately, this cannot be added to the components at the moment.

**Comment by acstll** [comment]
Hey @tomkroschwald thanks for opening this issue.

These components —button, link and all form elements— don't need `tabindex` set in order to be accessible because they already receive focus by default. You could only use `tabindex` on them for exactly the opposite, to block them from receiving focus, and that would then make them not accessible.

To support what I'm saying let me quote the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex). The possible values for `tabindex` are:

- a negative value (`tabindex=-1`): "means that the element is not reachable via sequential keyboard navigation, but could be focused with JavaScript"
- zero (`tabindex=0`): "means that the element should be focusable in sequential keyboard navigation, after any positive tabindex values and its order is defined by the document's source order." — **this is the default behaviour for `button` and `a` and `input`, etc.**
- a positive value: `tabindex=4`: "means the element should be focusable in sequential keyboard navigation, with its order defined by the value of the number." — **this is considered a bad practice, since it can break the sequential order for some users!**

For other components, with less or no semantics, you can indeed set `tabindex=0` on them, right in the custom element (because these are HTML elements after all) and they will receive focus as expected, e.g. `scale-card tabindex="0"` even if `tabindex` is not part of the component's listed attributes, because as I mentioned, all custom elements inherit from `HTMLElement`.

Lastly, components needing or requiring focus and keyboard support in order to work, like Tab Navigation or Modal, already handle this out-of-the-box. They have been AA-certified by an external team. So you probably don't need to set `tabindex` manually. A few components have a `focusable` boolean attribute, also.

It's worth mentioning that accessibility is not science: we (or the external team) can make mistakes, there could be differences between how different platforms or devices behave, or even different interpretations of the spec. So we're always eager for feedback and open to improve things wherever possible. Accessibility is very important to us. 🙂 

Is there any specific component where you actually need to set `tabindex` that is not working the way you expect?

**Comment by tomkroschwald** [comment]
> Hey @tomkroschwald thanks for opening this issue.
> 
> These components —button, link and all form elements— don't need `tabindex` set in order to be accessible because they already receive focus by default. You could only use `tabindex` on them for exactly the opposite, to block them from receiving focus, and that would then make them not accessible.
> 
> To support what I'm saying let me quote the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex). The possible values for `tabindex` are:
> 
> * a negative value (`tabindex=-1`): "means that the element is not reachable via sequential keyboard navigation, but could be focused with JavaScript"
> * zero (`tabindex=0`): "means that the element should be focusable in sequential keyboard navigation, after any positive tabindex values and its order is defined by the document's source order." — **this is the default behaviour for `button` and `a` and `input`, etc.**
> * a positive value: `tabindex=4`: "means the element should be focusable in sequential keyboard navigation, with its order defined by the value of the number." — **this is considered a bad practice, since it can break the sequential order for some users!**
> 
> For other components, with less or no semantics, you can indeed set `tabindex=0` on them, right in the custom element (because these are HTML elements after all) and they will receive focus as expected, e.g. `scale-card tabindex="0"` even if `tabindex` is not part of the component's listed attributes, because as I mentioned, all custom elements inherit from `HTMLElement`.
> 
> Lastly, components needing or requiring focus and keyboard support in order to work, like Tab Navigation or Modal, already handle this out-of-the-box. They have been AA-certified by an external team. So you probably don't need to set `tabindex` manually. A few components have a `focusable` boolean attribute, also.
> 
> It's worth mentioning that accessibility is not science: we (or the external team) can make mistakes, there could be differences between how different platforms or devices behave, or even different interpretations of the spec. So we're always eager for feedback and open to improve things wherever possible. Accessibility is very important to us. 🙂
> 
> Is there any specific component where you actually need to set `tabindex` that is not working the way you expect?

Hey @acstll,
thank you for your response.

In our case we need to change the `tabindex`. We have a table and need a custom tab order to fulfill ergonomic requirements our client has given to us.

**Comment by acstll** [comment]
Thanks for the quick reply @tomkroschwald 

As much as I would like to suggest to try and solve the custom tab order via markup and CSS, I find it's perfectly reasonable that you be able to set `tabindex` the way you want.

We could use a bit of feedback from you (not binding):

Adding a `tabindex` attribute —with this exact name— to the components is problematic because then you would have the custom element (host) _and_ the inner element (button or link) both with `tabindex` set, which will yield an extra tab you don't want. I see two options for solving this:

- (a) we use an attribute with a different name, like `button-tabindex` or `inner-tabindex` (we've done this with `input-id` in `scale-text-field` for instance)
- (b) we don't add any attributes, and look for `tabindex` being set on e.g. `scale-button`; if set, we take the value and set `tabindex` in the inner button element, and remove the `tabindex` attribute from `scale-button`

Do you have a preference for one of these two options?

**Comment by tomkroschwald** [comment]
We would suggest option A, it's easier for debugging purposes. Thank you for your reply!

**Comment by acstll** [comment]
We have merged https://github.com/telekom/scale/pull/686 that adds an `inner-tabindex`attribute to `scale-button` and `scale-link`. This will go out in the next beta release. We will let you know.

**Comment by acstll** [comment]
The addition should be out in the latest beta, please try it out :)
